### PR TITLE
ft: add the "unreachable" analyzer

### PR DESCRIPTION
**Pull request by maurocchi**
## Summary

- add `paths` configuration option to report only functions defined in such paths (i.e. to hide code made only for tests purposes);
- add `limit` configuration option to report only a limited number of functions;
- make the `checks` options overridable, so that the user can enable the "unreachable" analyzer;
- add the `MixUnused.Analyzers.Unreachable` analyzer, alternative to the `Unused` one;
- fix usage detection of public functions called with default arguments;
- detect generated functions and omit them from report;
- log debug information if the `MIX_UNUSED_DEBUG` env var is set to `true`.

All the changes are backward compatible.

## The unused analyzer

It tries to detect all used code by performing a tricky depth-first visit on the calls graph, but we found cases where it fails to detect functions used transitively from exported functions.
Since the solution seems to require a different approach, we defined a new experimental `Unreachable` analyzer in order to avoid breaking changes in the `Unused` one.

## The unreachable analyzer

It just analyses the calls graph finding all the functions that are reachable from a specific set of well-known used functions. This makes the analysis simple and deterministic.

The analyzer requires the user to specify the set of usages, or "usages-discovery" (modules that try to discover dynamically-called functions - for instance by reading the application config or by analyze the source code).

`Unreachable` has some differences from `Unused`:
* Unreachable does not assume that all behaviour implementers are used by default.
* Unreachable does not report generated functions.
* Unreachable, by default, does not report unused functions called transitively (to make the report cleaner).

Please look at the `README.md` and `guides/unreachable-analyzer.md` for further details.

### CODE REVIEW

File: lib/mix_unused/analyzers/unreachable/entrypoints/amqpx_consumers_discovery.ex [hunk 1]
```diff
@@ -0,0 +1,20 @@
+defmodule MixUnused.Analyzers.Unreachable.Entrypoints.AmqpxConsumersDiscovery do
+  @moduledoc """
+  Discovers the consumers configured for the amqpx library.
```
Comment (net02) [inline lib/mix_unused/analyzers/unreachable/entrypoints/amqpx_consumers_discovery.ex]:
```suggestion
  Discovers the consumers configured for the [amqpx library](https://hex.pm/packages/amqpx).
```

File: lib/mix_unused/analyzers/unreachable/entrypoints/http_mock_pal_discovery.ex [hunk 2]
```diff
@@ -0,0 +1,14 @@
+defmodule MixUnused.Analyzers.Unreachable.Entrypoints.HttpMockPalDiscovery do
+  @moduledoc """
+  Discovers the mock modules configured for the http_mock_pal library.
```
Comment (net02) [inline lib/mix_unused/analyzers/unreachable/entrypoints/http_mock_pal_discovery.ex]:
```suggestion
  Discovers the mock modules configured for the [http_mock_pal library](https://hex.pm/packages/http_mock_pal).
```

File: lib/mix_unused/analyzers/unreachable.ex [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  @impl true
+  def analyze(data, exports, config) do
+    config = Config.cast(config)
+    graph = Calls.calls_graph(data, exports)
+    usages = Usages.usages(config, exports)
+    reachables = graph |> Graph.reachable(usages) |> MapSet.new()
+    called_at_compile_time = Calls.called_at_compile_time(data, exports)
+
+    for {mfa, _meta} = call <- exports,
+        filter_transitive_call?(config, graph, mfa),
```
Comment (succoDiPompelmo) [inline lib/mix_unused/analyzers/unreachable.ex:28]: Penso di aver capito il giro del fumo su questo tool, ma alcune zone sono rimaste oscure, tipo questa come mai abbiamo bisogno di trovare le transitive call ?

**Review by succoDiPompelmo — APPROVED** [review]
Non vedo niente di strano. Unico grande down è l'assenza di documentazione che integrerei direttamente nel README, visto anche la complessità del tool. Poi sarebbe anche carino fare un piccolo workshop interno per "spiegare" il tool.

File: lib/mix_unused/analyzers/unreachable.ex (hunk 3 above)
Comment (korialis) [reply lib/mix_unused/analyzers/unreachable.ex:28]: questo è un filtering che facciamo in base all'opzione `report_transitively_unused` per fare il report delle sole radici del grafo delle chiamate che potrebbero risultare non utilizzate.  Il tool altrimenti farebbe il report di ogni chiamata non "reachable". Serve a facilitare il processo di individuazione degli entrypoint

**Comment by korialis** [comment]
Code looks clean and correct to me :)
In my PR #3 I applied two minor edits (in `Aliases.resolve/2` and `SupervisorDiscovery.discover_usages/1`)

**Review by korialis — APPROVED** [review]

**Comment by mattiabasone** [comment]
> Poi sarebbe anche carino fare un piccolo workshop interno per "spiegare" il tool.

:+1:

**Review by mbusi — COMMENTED** [review]
wow cool; nothing to add as a reviewer, sadly; just a quick question: why the fork? do you think it's a custom approach that couldn't be used in the public libray?

really gj btw <3

**Comment by maurocchi** [comment]
> why the fork? do you think it's a custom approach that couldn't be used in the public library?

Thanks @mbusi, of course the fork is to modify the open source repo 🙂
I think we could share our changes when they will be stable.

File: lib/mix_unused/filter.ex [hunk 4]
```diff
@@ -24,23 +24,35 @@ defmodule MixUnused.Filter do
...
-        {m, f} -> {m, f, :_}
-        {m} -> {m, :_, :_}
-        m when is_atom(m) -> {m, :_, :_}
-        %Regex{} = m -> {m, :_, :_}
-        cb when is_function(cb) -> cb
-      end)
-
-    Enum.reject(exports, fn {func, meta} ->
-      Enum.any?(filters, &mfa_match?(&1, func, meta))
+    Map.reject(exports, matcher(patterns))
```
Comment (net02) [inline lib/mix_unused/filter.ex:27]: questa ed un paio di `Map.filter` qui sotto ed in un altro file andrebbero cambiati per mantenere la compatibilità con elixir ~1.10 (dichiarata dalla libreria), visto che sono stati aggiunti nella 1.13

```suggestion
   exports |> Enum.reject(matcher(patterns)) |> Enum.into(%{})
```

**Review by net02 — CHANGES_REQUESTED** [review]
vedendo [la PR](https://github.com/hauleth/mix_unused/pull/37) c'è anche da rebasare il titolo di un commit che non piace alle sue github actions 😉

**Comment by maurocchi** [comment]
Close as it has already been squashed.
We keep this branch alive because there are some projects that refer to it.

**Pull request closed without merging**